lib/logstash/outputs/s3.rb

```diff
@@ -106,6 +109,13 @@ class LogStash::Outputs::S3 < LogStash::Outputs::Base
   # S3 bucket
   config :bucket, :validate => :string, :required => true
 
+  # Specify a custom endpoint for use with non-AWS S3 implementations, e.g.,
+  # Ceph.  Provide a URL in the format http://127.0.0.1:8080/
+  config :endpoint, :validate => :string
+
+  # When false, specify the bucket in the subdomain.  When true, specify the bucket in the path.
+  config :force_path_style, :validate => :bool, :default => false
```

Comment:
I do agree the naming is terrible, but it's inherited from the AWS documentation.

There are two inputs: S3-server-address (`host+port`), bucket.

There are two common formats:
* Path: `http://host:port/bucket/`, `https://host:port/bucket/`
** AWS documentation uses: _Path-style_ in AWS documentation.
* Subdomain: `http://bucket.host:port/`, `https://bucket.host:port/`
** AWS documentation uses both: _Subdomain_ calling format and _Virtual Hosted–Style_
* (and some more exotic variations in calling format that are not relevant to us)

To use the Subdomain format, there are a LOT of conditions:
* If there is a DNS wildcard for `*.host` AND the bucket name is DNS-compatible [lowercase-only, no characters invalid in DNS]
* If there is SSL on the port [which might NOT be port 443], there ALSO needs to be a wildcard SSL certificate and the bucket name ALSO must not contain periods [which would cause the cert to never match [or you have to write custom HTTPS hostname validation].

If you're running something like minio or a small ceph install as a local S3 server, and thus have an IP+port only for connecting, eg http://127.0.0.1:8080/, then you MUST use path-style access. You cannot use Subdomain access at all. This is also relevant if you don't have a DNS wildcard.

if you're connecting to a larger S3 implementation, including AWS & multi-region/federated Ceph deployments, and the protocol ALLOWS you to be issued a 307 redirection to another host, usually one where your data is actually sorted (this can be seen on AWS if you create an EU s3 bucket, then try to access it via the US endpoints): this redirect can add significant latency in some cases, so hitting the correct host right away is important, and is more common in path-style access than subdomain.